chapters/annotations.tex

```diff
@@ -336,11 +336,17 @@ \section{Annotations for Code Generation}\label{annotations-for-code-generation}
 "Evaluate" "=" ( false | true )
 \end{lstlisting}\end{synopsis}
 \begin{semantics}
-The annotation \lstinline!Evaluate! can occur in the component declaration, its type declaration, or a base-class of the type-declaration.  In the case of multiple conflicting annotations it is handled similarly to modifiers (e.g., an \lstinline!Evaluate! annotation on the component declaration takes precedence).  The annotation \lstinline!Evaluate! only has effect for a component declared with the prefix \lstinline!parameter!.
+The annotation \lstinline!Evaluate! can occur in the component declaration, its type declaration, or a base-class of the type-declaration.
+In the case of multiple conflicting annotations it is handled similarly to modifiers (e.g., an \lstinline!Evaluate! annotation on the component declaration takes precedence).
+The annotation \lstinline!Evaluate! is only allowed for parameters and constants.
 
-If \lstinline!Evaluate = true!, the model developer proposes to utilize the value for the symbolic processing. In that case, it is not possible to change the parameter value after symbolic pre-processing.
+Setting \lstinline!Evaluate = true! is only allowed for structural parameters and constants.
+For a structural parameter, it means that its value must be determined during translation, similar to a constant.
```

Comment:
```suggestion
```
This sentence ia also  a lttle too strong and can be omitted.

Reply:
But if it only means that the parameter must be structural, then it's almost pointless.  Now that it has been ensured that it can be evaluated, I don't really see the reasons why we wouldn't go all the way and require that it is determined during translation.  Do you have some proposal for what to say instead to make it mean something more than just the explicit absence of `Evaluate = false`?

Reply:
Why do you want to force evaluation of parameters? If they are not presented as such at the end of analysis (and structural parameters shouldn't), it's not observable, whether the parameters are evaluated or not and it does not change the results of the simulation. Thats internal to the compiler. The only thing is to allow the compiler the evaluation and thats already said in the next sentence.

Reply:
Wait, could this be a misunderstanding?  Where it says _For a structural parameter_ here, it gives meaning to `Evaluate = true` in the context of a structural parameter, not to structural parameters in general.  Can we clarify this before going back to your comment above?

Reply:
For me `Evaluate=true` allows for evaluation of parameters which are only used in contexts which do not force an evaluation, e.g. in conditions of if/then/else constructs with equal number of branches. So not used branches can be removed and therefore strange equation systems are avoided. It is some kind of optimization by the modeler.
So in this case they become structural parameters by putting the annotation there. So the question is, what ist first, the annotation or the propertiy of structural parameter. The parameter doesn't have to be structural as long as there is no such annotation (and no such usage). But if it is used in contexts, which forces them to be structural, then the annotation is not needed.

Reply:
It think what you're saying is this, here expressed using the proposed terminology:
> The parameter doesn't have to be -structural- _evaluated_ as long as there is no such annotation (and no such usage).

In the proposed terminology, being a structural parameter has to do with variability analysis; it is decoupled from how the parameter is used.  Depending on how it is used, it may become evaluated ("determined during translation").

When it comes to `Evaluate`, only `Evaluate = false` has an impact on classification of parameters as normal or structural; setting `Evaluate = true` cannot change the classification, but is an error if the parameter isn't structural.  Hence, one possible use of `Evaluate = true` is to _demonstrate_ that a parameter is structural.

> For me `Evaluate=true` allows for evaluation of parameters which are only used in contexts which do not force an evaluation, e.g. in conditions of if/then/else constructs with equal size of branches. So not used branches can be removed and therefore strange equation systems are avoided. It is some kind of optimization by the modeler.
So in this case they become structural parameters by putting the annotation there.

This sounds similar to the use of `Evaluate = true` I have in mind.  The difference, I think, is that the proposal is to _allow_ such evaluation for _all_ structural parameters (but I'd prefer a tool which tries hard to avoid it), and enable users to _force_ such evaluation by setting `Evaluate = true`.

Reply:
So, as I understand, being structural parameter and beeing evaluated is something different. Usually parameters fulfill almost always the requirements for structural parameters, but you never want to evaluate them, if it is not really forced. Even if they are not evaluated, you will call them structural parameters. (Is this right?).
There are two ways to force it, either by using them in a context which needs evaluation or by annotation `Evaluate = true`.
Is this the right understanding? If yes, the only difference is, that I see the annoation not that strong, because in the current specification the "developer proposes to utilize the value for symbolic processing". If it is not necessary its not done. 
But, maybe that is what you want to change, because the modeler does not know what he can expect from the compiler in that case, whether it is necessary or not. It is much more clearer to say, if the annotation is there, the parameter is evaluated and if it is not possible it is an error (even of it is not really necessary - annotations normally do not cause errors).
So I would be in favor of the proposal, if there wouldn't be the additional errors.

Reply:
> But, maybe that is what you want to change, because the modeler does not know what he can expect from the compiler in that case, whether it is necessary or not. It is much more clearer to say, if the annotation is there, the parameter is evaluated and if it is not possible it is an error (even of it is not really necessary - annotations normally do not cause errors).

Except, that there are cases where you want to ignore `Evaluate=true`.
A typical use-case is that a component contains `Evaluate=true` for efficiency (there are a number of cases in MSL) - but you want to do a parameter study, and thus want to be able to change that parameter .The simple solution is to bind the parameter to another parameter having `Evaluate=false`, and give that annotation higher priority - and not treat it as an error.

Reply:
> So, as I understand, being structural parameter and beeing evaluated is something different. Usually parameters fulfill almost always the requirements for structural parameters, but you never want to evaluate them, if it is not really forced. Even if they are not evaluated, you will call them structural parameters. (Is this right?).

Yes, and I share your desire for them not becoming evaluated – I'll keep this in mind when selecting my Modelica tool.

> There are two ways to force it, either by using them in a context which needs evaluation or by annotation `Evaluate = true`.
> Is this the right understanding?

No, for two reasons:
- I didn't want to step on the toes of tools that sometimes do more aggressive evaluation of parameters.
- I don't want to make assumptions about exactly what a tool might need to evaluate during translation, so this is left as a quality of implementation.  In general, I think it would be OK for something to be a structural expression in the specification, as long as we find it reasonable to support tools that need translation-time evaluation for this expression; tools that are able to avoid translation time evaluation are still free to determine the value at initialization instead.

> If yes, the only difference is, that I see the annoation not that strong, because in the current specification the "developer proposes to utilize the value for symbolic processing". If it is not necessary its not done.
> But, maybe that is what you want to change, because the modeler does not know what he can expect from the compiler in that case, whether it is necessary or not. It is much more clearer to say, if the annotation is there, the parameter is evaluated and if it is not possible it is an error (even of it is not really necessary - annotations normally do not cause errors).

Yes, there is a change in nuance here and it is good that you brought it up for discussion.

> So I would be in favor of the proposal, if there wouldn't be the additional errors.

Glad to hear!

Reply:
> Except, that there are cases where you want to ignore `Evaluate=true`.
> A typical use-case is that a component contains `Evaluate=true` for efficiency (there are a number of cases in MSL) - but you want to do a parameter study, and thus want to be able to change that parameter .The simple solution is to bind the parameter to another parameter having `Evaluate=false`, and give that annotation higher priority - and not treat it as an error.

I interpret _simple solution_ as a suggestion for how to modify the proposal, and will give this a try shortly.

In the meantime, why bind it to a parameter with `Evaluate = false` rather than just removing `Evaluate = true`?  (The answer can probably be used as an illustrative example.)

Reply:
Because the Evaluate=true is found in a library component - like Modelica.Mechanics.MultiBody, and you cannot easily change that - and even if you could it would be time-consuming.

Reply:
Would you find it acceptable to issue a warning in case `Evaluate = true` is used on a normal parameter?  Then a high quality warning mechanism will be able to act intelligently depending on whether there's an `Evaluate = false` behind normal variability of the declaration equation, or if the parameter is normal for other reasons.

Reply:
The rules for `Evaluate` have been updated now to handle this use case, illustrated by an example.

Reply:
This conversation is very old and outdated in view of the last design meeting discussions.  Resolving.